lab04 - numeronyms

n0npax opened the pull request:
Implement numeronyms
Implement unit tests

Fixes #210 

Review of colleague's PR #197

#### Changes proposed in this PR:

## - Solve lab 04
implement main logic
implement unit tests

n0npax committed:
Completed lab04 - numeronyms

Implement numeronyms
Implement unit tests

nicholascross commented on 04_numeronym/n0npax/main.go:
```diff
@@ -0,0 +1,29 @@
+var out io.Writer = os.Stdout
+
+func shorten(word string) string {
+	if len(word) < 4 {
```
Could potentially use a local variable for length I notice it is derived several times.

n0npax replied:
Thanks for this. Fixed

nicholascross commented on 04_numeronym/n0npax/main_test.go:
```diff
@@ -0,0 +1,54 @@
+		"internationalization": {input: "internationalization", expected: "i18n"},
+	}
+	for name, test := range testCases {
+		t.Run(name, helperTestShorten(test.input, test.expected))
```
Could we just use the anonymous function inline?  I don't believe we would lose any readability, it might make it a bit more succinct.

n0npax replied:
It will cause referring to variables out of the scope and `golangci-lint` will complain

nicholascross replied:
I believe it is probably complaining because the function is capture the test value which will possibly be incorrect by the time the function is executed.  You can resolve this by shadowing the variable.

```go
for name, test := range cases {
		input := test.input
		want := test.want
		t.Run(name, func(t *testing.T) {
			assert.Equalf(t, want, numeronyms(input...), "%v was not sorted", input)
		})
	}
```

~If you really want to use a separate function do we still need the internal anonymous function inside `helperTestShorten`?  Is it possible to use it directly?~

**EDIT**

I think if you tried to use it directly you would probably end up with the same problem again, so maybe that is not a good suggestion.  Separating out into a separate function achieves the same thing as creating local variables it is just more verbose.

n0npax replied:
The suggested solution works. Thanks!
I'm surprised the shadowing solves this. `want` and `input` is still outside of the anonymous function.

n0npax committed:
CR fixes

Store len in wariable istrad of invoking function many times
Shadow variables in

anzboi commented on 04_numeronym/n0npax/main.go:
```diff
@@ -0,0 +1,30 @@
+
+var out io.Writer = os.Stdout
+
+func shorten(word string) string {
```
Don't feel the need to change this, since this is already fine as is.
Just wanted to point out you can also assign `wordlen` in the `if` block...
```
if wordlen := len(word); wordlen < 4 {
  return word
} else {
  return fmt.Sprintf("%c%d%c", word[0], wordlen-2, word[wordLen-1])
}
```

n0npax replied:
thanks for hint

anzboi commented on 04_numeronym/n0npax/main_test.go:
```diff
@@ -0,0 +1,51 @@
+		input, want := test.input, test.expected
+		t.Run(name, func(t *testing.T) {
+			actual := shorten(input)
+			assert.Equalf(t, want, actual, "For input: %v expected: %v got %v", input, want, actual)
```
Again, don't feel the need to change this.
You can avoid the need to pass `t` into these `testify` functions by creating an assertions object at the start of the test
```
func TestName(t *testing.T) {
  a := assert.New(t)
  // Test Code
  a.Equalf(want, actual, "message") // Don't need to pass t
}
```

This is really a matter of preference, you trade having to pass `t` for having to create `a` at the start.

nicholascross replied:
Thanks for pointing this out 👍

anzboi reviewed (approved):
LGTM
Some comments on how you can do things a little differently, but these are mostly up to preference.
